### PR TITLE
router: vhost/route/w.cluster local filter configuration

### PR DESCRIPTION
**Pull request by rodaine**
This patch adds support for resolving virtual host-, route-, and weighted cluster-local configuration for filters. See https://github.com/envoyproxy/data-plane-api/issues/540 for discussion. Configs are provided via RDS as opaque structs (similar to the existing LDS-provided configs), resolved to concrete message types only once, and exposed by the VirtualHost, Route and RouteEntry types. These RDS-scoped configs can be a different pb message type than the base configuration, but default to being identical. Filters can optionally access these configs to change their behavior accordingly.

I will be following up to this PR with making the buffer filter route-aware. @rshriram will also be working on the fault and lua filters to support this.

*Risk Level*: Low-Medium
*Testing*: unit
*API/Docs Changes*: https://github.com/envoyproxy/data-plane-api/pull/605 
*Release Notes*: (will update in this PR)

### CODE REVIEW

**Comment by rshriram** [comment]
cc @lizan @qiwzhang

File: include/envoy/server/filter_config.h [hunk 1]
```diff
@@ -295,6 +295,15 @@ class NamedHttpFilterConfigFactory {
...
   virtual ProtobufTypes::MessagePtr createEmptyConfigProto() { return nullptr; }
 
+  /**
+   * @return ProtobufTypes::MessagePtr create an empty virtual host, route, or weight cluster-local
+   *         config proto message for v2. The filter config, which arrives in an opaque
+   *         google.protobuf.Struct message, will be converted to JSON and then parsed into this
+   *         empty proto. By default, this method returns the same value as createEmptyConfigProto,
+   *         and can be optionally overridden in implementations.
+   */
+  virtual ProtobufTypes::MessagePtr createEmptyRDSConfigProto() { return createEmptyConfigProto(); }
```
Comment (lizan) [inline include/envoy/server/filter_config.h]: s/RDS/Route/ ?
The route config can come from static config as well.

File: bazel/repository_locations.bzl [hunk 2]
```diff
@@ -75,8 +63,12 @@ REPOSITORY_LOCATIONS = dict(
         strip_prefix = "protobuf-3.5.0",
         urls = ["https://github.com/google/protobuf/archive/v3.5.0.tar.gz"],
     ),
+    com_lightstep_tracer_cpp = dict(
+        commit = "6a198acd328f976984699f7272bbec7c8b220f65",
+        remote = "https://github.com/lightstep/lightstep-tracer-cpp",  # v0.6.1
+    ),
     envoy_api = dict(
-        commit = "2dcc435e8ae1d35f8c3a4fa9f132778482fb1a78",
+        commit = "c15104fce713856203fa4ad948d752e9a1f657c1",
```
Comment (lizan) [inline bazel/repository_locations.bzl]: revert this file other than this line? those seems unrelated.

File: source/common/config/utility.h [hunk 3]
```diff
@@ -219,6 +219,20 @@ class Utility {
     return config;
   }
 
+  template <class ProtoMessage, class Factory>
```
Comment (lizan) [inline source/common/config/utility.h]: I don't think you need the first template parameter, it is always Struct.

File: source/common/router/config_impl.cc [hunk 4]
```diff
@@ -854,5 +876,27 @@ ConfigImpl::ConfigImpl(const envoy::api::v2::RouteConfiguration& config, Runtime
...
 
+PerFilterConfigs::PerFilterConfigs(const Protobuf::Map<std::string, ProtobufWkt::Struct>& configs) {
+  for (const auto& cfg : configs) {
+    const std::string& name = cfg.first;
+    const ProtobufWkt::Struct& struct_config = cfg.second;
+
+    auto& factory = Envoy::Config::Utility::getAndCheckFactory<
+        Server::Configuration::NamedHttpFilterConfigFactory>(name);
+
+    configs_[name] = Envoy::Config::Utility::translateToFactoryRDSConfig(struct_config, factory);
```
Comment (lizan) [inline source/common/router/config_impl.cc]: Need to perform validation when PerFilterConfigs are loaded too.

File: include/envoy/router/router.h [hunk 5]
```diff
@@ -258,6 +258,13 @@ class VirtualHost {
    * @return const Config& the RouteConfiguration that owns this virtual host.
    */
   virtual const Config& routeConfig() const PURE;
+
+  /**
+   * @return const Protobuf::Message* the per-filter config for the given
+   * filter name configured for this virtual host. If none is present, the
```
Comment (mattklein123) [inline include/envoy/router/router.h]: super nit: "the nullptr" is not typically said. Just "nullptr." Same below.

File: include/envoy/server/filter_config.h (hunk 1 above)
Comment (mattklein123) [reply include/envoy/server/filter_config.h]: +1

File: source/common/config/utility.h [hunk 6]
```diff
@@ -219,6 +219,20 @@ class Utility {
...
+  static ProtobufTypes::MessagePtr translateToFactoryRDSConfig(const ProtoMessage& source,
+                                                               Factory& factory) {
+    ProtobufTypes::MessagePtr config = factory.createEmptyRDSConfigProto();
+
+    if (config == nullptr) {
```
Comment (mattklein123) [inline source/common/config/utility.h]: I would just RELEASE_ASSERT() this like I recently did for the other instances of this.

File: source/common/router/BUILD [hunk 7]
```diff
@@ -22,6 +22,7 @@ envoy_cc_library(
         "//include/envoy/http:header_map_interface",
         "//include/envoy/router:router_interface",
         "//include/envoy/runtime:runtime_interface",
+        "//include/envoy/server:filter_config_interface",
```
Comment (mattklein123) [inline source/common/router/BUILD]: In general we try to not have common depend on server, though this isn't really enforced explicitly. I would add a TODO here and in code to figure out a better way to break this cycle.

File: source/common/router/config_impl.cc (hunk 4 above)
Comment (mattklein123) [reply source/common/router/config_impl.cc]: I don't think we can do this unfortunately. What type would we validate? (This is related to me asking very nicely for `Validate()` to be a method on the base proto message which is virtual and empty and can be overriden...

File: test/common/router/config_impl_test.cc [hunk 8]
```diff
@@ -4201,6 +4207,163 @@ name: foo
...
+  nanos: 789
+)EOF";
+
+  ProtobufWkt::Struct buffer_cfg;
+  MessageUtil::loadFromYaml(yaml, buffer_cfg);
+  Protobuf::Map<std::string, ProtobufWkt::Struct> proto_cfgs;
+  proto_cfgs[factory.name()] = buffer_cfg;
+  PerFilterConfigs configs{proto_cfgs};
+
+  ASSERT_EQ(nullptr, configs.get("unknown.filter"))
```
Comment (mattklein123) [inline test/common/router/config_impl_test.cc]: s/ASSERT/EXPECT in all places for the most part.

File: test/common/router/config_impl_test.cc [hunk 9]
```diff
@@ -13,10 +16,13 @@
 #include "common/network/address_impl.h"
 #include "common/router/config_impl.h"
 
+#include "server/config/http/buffer.h"
```
Comment (mattklein123) [inline test/common/router/config_impl_test.cc]: This has been moved and will need a master merge.

File: test/common/router/config_impl_test.cc [hunk 10]
```diff
@@ -4201,6 +4207,163 @@ name: foo
...
+  ASSERT_NO_THROW({
+    const auto buffer = dynamic_cast<const envoy::config::filter::http::buffer::v2::Buffer*>(cfg);
+    ASSERT_EQ(expected_max_bytes, buffer->max_request_bytes().value())
+        << "config value does not match expected for source: " << source;
+  }) << "config should properly dynamic_cast to the appropriate type for source: "
+     << source;
+}
+
+TEST(PerFilterConfigsTest, RouteLocalConfig) {
+  Server::Configuration::BufferFilterConfig factory;
```
Comment (mattklein123) [inline test/common/router/config_impl_test.cc]: Instead of using buffer filter, can you register a test filter for this test, so that you can correctly verify that you can override to be a different type for the filter data? If you need to you can have a proto just for testing.

File: source/common/router/config_impl.cc (hunk 4 above)
Comment (mattklein123) [reply source/common/router/config_impl.cc]: Pre-processing is being treated as an additional feature. That will need more thinking. We won't be doing that as part of this PR.

File: bazel/repository_locations.bzl (hunk 2 above)
Comment (rodaine) [reply bazel/repository_locations.bzl]: the vscode bazel plugin seems a bit overzealous in its linting...

File: source/common/config/utility.h (hunk 3 above)
Comment (rodaine) [reply source/common/config/utility.h]: was being consistent with the other translate function but will update!

File: source/common/config/utility.h (hunk 6 above)
Comment (rodaine) [reply source/common/config/utility.h]: I was being consistent with the translate method above. Shall I update both?

File: source/common/router/BUILD (hunk 7 above)
Comment (rodaine) [reply source/common/router/BUILD]: sg. i imagine this could be done with bazel visibility rules?

File: source/common/config/utility.h (hunk 6 above)
Comment (mattklein123) [reply source/common/config/utility.h]: Yes please. I don't think it's really needed to throw an exception here.

File: source/common/router/BUILD (hunk 7 above)
Comment (mattklein123) [reply source/common/router/BUILD]: I need to think about it. Will look at it as part of repo reorg work.

File: test/common/router/config_impl_test.cc (hunk 10 above)
Comment (rodaine) [reply test/common/router/config_impl_test.cc]: is there an existing one I can use for this or should I create one? Should I create test proto in envoy or data plane?

File: include/envoy/router/router.h [hunk 11]
```diff
@@ -258,6 +258,13 @@ class VirtualHost {
    * @return const Config& the RouteConfiguration that owns this virtual host.
    */
   virtual const Config& routeConfig() const PURE;
+
+  /**
+   * @return const Protobuf::Message* the per-filter config for the given
+   * filter name configured for this virtual host. If none is present,
+   * nullptr is returned.
+   */
+  virtual const Protobuf::Message* perFilterConfig(const std::string& name) const PURE;
```
Comment (qiwzhang) [inline include/envoy/router/router.h]: This suggestion is trying to address this issue:  https://github.com/envoyproxy/envoy/issues/3008
We can do it in a separate PR.  But it is so simple, you can also put in in this PR too.

In these VHost/Route/Cluster,   add another function

// If perFitlerConfig exists for this filter, get its config hash value to detect if config is updated.
// return 0 if no config for the filter.
size_t perFitlerConfigHash(const std::string& name) const PURE;

File: source/common/router/config_impl.cc [hunk 12]
```diff
@@ -854,5 +876,27 @@ ConfigImpl::ConfigImpl(const envoy::api::v2::RouteConfiguration& config, Runtime
...
 
+PerFilterConfigs::PerFilterConfigs(const Protobuf::Map<std::string, ProtobufWkt::Struct>& configs) {
+  for (const auto& cfg : configs) {
+    const std::string& name = cfg.first;
+    const ProtobufWkt::Struct& struct_config = cfg.second;
+
+    auto& factory = Envoy::Config::Utility::getAndCheckFactory<
+        Server::Configuration::NamedHttpFilterConfigFactory>(name);
+
+    configs_[name] = Envoy::Config::Utility::translateToFactoryRouteConfig(struct_config, factory);
```
Comment (qiwzhang) [inline source/common/router/config_impl.cc]: create another map 

config_hash_ to store hash of config

File: include/envoy/router/router.h (hunk 11 above)
Comment (mattklein123) [reply include/envoy/router/router.h]: I don't quite understand how this will help you. In the filter are you going to check this on a per-route basis? What you really want is another callback which allows a filter to update global structures on a route update.

File: source/common/router/config_impl.cc (hunk 12 above)
Comment (mattklein123) [reply source/common/router/config_impl.cc]: Same comment as above. I don't think this is very helpful and I would rather sort out filter update callbacks as a separate feature?

File: test/common/router/config_impl_test.cc (hunk 10 above)
Comment (mattklein123) [reply test/common/router/config_impl_test.cc]: Just make one local to the test.

File: include/envoy/router/router.h (hunk 11 above)
Comment (qiwzhang) [reply include/envoy/router/router.h]: This is to help filters need to pre-process per_route_config.  per_route_config can be dynamically changed anytime.  But you don't want to pre-process the config for each request.  This hash will help to detect config change, and skip pre_processing step:

This is how it can be used,  For each request in decodeHeader()

    auto hash = route()->perFilterConfigHash("my_filter_name");
    if (hash == 0) // handle not per_route_config case

     obj =  global_obj_map.find(hash);
     if (obj == null) {
             obj = pre_process_config( route()->perFilterConfig("my_filter_name") );
             global_obj_map[hash] = obj;
     }

This will also achieve:  different routes with same config can share the same pre_processed object.

Comment (mattklein123) [reply include/envoy/router/router.h]: This is not how I would recommend implementing this. (The above is not thread safe.). I have a pretty good idea of how to efficiently implement this and give you the functionality you need. Let's do this in a follow up.

Comment (qiwzhang) [reply include/envoy/router/router.h]: In istio mixer-client filter,  the "global_map" actually is per_thread_global_map.  so this approach works for us. 
I agreed that the filter_factory callback approach is better, the global_map can be used by all worker threads. Let us do this in a followup PR.
Istio is waiting on this feature.   If you don't have anybody to work on this, I can help.

**Comment by rodaine** [comment]
Ok, changes made. PTAL @mattklein123

File: test/common/router/config_impl_test.cc [hunk 13]
```diff
@@ -4204,6 +4209,148 @@ name: foo
...
+  std::string yaml = R"EOF(
+foo: "bar"
+)EOF";
+
+  ProtobufWkt::Struct some_cfg;
+  MessageUtil::loadFromYaml(yaml, some_cfg);
+  Protobuf::Map<std::string, ProtobufWkt::Struct> proto_cfgs;
+  proto_cfgs["unknown.filter"] = some_cfg;
+
+  EXPECT_THROW(PerFilterConfigs{proto_cfgs}, EnvoyException)
```
Comment (mattklein123) [inline test/common/router/config_impl_test.cc]: I would use EXPECT_THROW_WITH_MESSAGE here for better checking. Same elsewhere if applicable.

File: test/common/router/config_impl_test.cc [hunk 14]
```diff
@@ -4204,6 +4209,148 @@ name: foo
...
+    EXPECT_NE(nullptr, cfg) << "config should not be null for source: " << source;
+    EXPECT_NO_THROW({
+      const auto ts = dynamic_cast<const ProtobufWkt::Timestamp*>(cfg);
+      EXPECT_EQ(expected_seconds, ts->seconds())
+          << "config value does not match expected for source: " << source;
+    }) << "config should properly dynamic_cast to the appropriate type for source: "
+       << source;
+  }
+
+  std::unique_ptr<Registry::InjectFactory<Server::Configuration::NamedHttpFilterConfigFactory>>
```
Comment (mattklein123) [inline test/common/router/config_impl_test.cc]: nit: I think this can just be stack allocated with the test? Do you need a unique_ptr?

File: test/common/router/config_impl_test.cc [hunk 15]
```diff
@@ -4204,6 +4209,148 @@ name: foo
...
+  EXPECT_NE(nullptr, processed) << "filter configs present should return a concrete Message";
+
+  EXPECT_NO_THROW({
+    const ProtobufWkt::Timestamp* cfg = dynamic_cast<const ProtobufWkt::Timestamp*>(processed);
+    EXPECT_EQ(123, cfg->seconds());
+    EXPECT_EQ(456, cfg->nanos());
+  }) << "returned message should be castable to known type";
+}
+
+TEST_F(PerFilterConfigsTest, UnknownFilter) {
```
Comment (mattklein123) [inline test/common/router/config_impl_test.cc]: nit: This would be a better test if you had a full route config that got parsed that then resulted in the exception.

File: source/common/router/config_impl.cc [hunk 16]
```diff
@@ -855,5 +878,27 @@ ConfigImpl::ConfigImpl(const envoy::api::v2::RouteConfiguration& config, Runtime
...
+  }
+}
+
+const Protobuf::Message* PerFilterConfigs::get(const std::string& name) const {
+  auto cfg = configs_.find(name);
+
+  if (cfg == configs_.end()) {
```
Comment (mattklein123) [inline source/common/router/config_impl.cc]: nit: I would just use the ternary operator here to collapse the rest of these lines into one.

**Comment by zuercher** [comment]
lgtm, modulo pending comments.
